File: tests/workflow/consumer/rust/src/test/test_request_groupb_groupc_structb.rs
use super::{controller, samples, stat, ClientError, Consumer, StatEvent};
use tokio::sync::mpsc::UnboundedSender;

pub async fn execute(
    consumer: &mut Consumer<ClientError>,
    tx_stat: &UnboundedSender<StatEvent>,
) -> Result<(), String> {
    let mut case_b = false;
    let mut case_d = false;
    let mut case_c = false;
    let mut struct_a = false;
    while !struct_a || !case_b || !case_d || !case_c {
        let response = consumer
            .groupb_groupc_structb(samples::group_b::group_c::struct_b::get())
            .await
            .map_err(|e| e.to_string())?;
        tx_stat
            .send(StatEvent::Inc(stat::Alias::TestRequestGroupBGroupCStructB))
            .map_err(|e| e.to_string())?;
        match response {
            controller::GroupBGroupCStructBResponse::CaseB(res) => {
                if !samples::struct_b::equal(res.clone()) {
                    return Err(format!("Invalid data received: {:?}", res));
                }
                case_b = true;
            }
            controller::GroupBGroupCStructBResponse::CaseC(res) => {
                if !samples::struct_c::equal(res.clone()) {
                    return Err(format!("Invalid data received: {:?}", res));
                }
                case_c = true;
            }
            controller::GroupBGroupCStructBResponse::CaseD(res) => {
                if !samples::struct_d::equal(res.clone()) {
                    return Err(format!("Invalid data received: {:?}", res));
                }
                case_d = true;
            }
            controller::GroupBGroupCStructBResponse::Err(res) => {
                if !samples::group_b::group_c::struct_a::equal(res.clone()) {
                    return Err(format!("Invalid data received: {:?}", res));
                }
                struct_a = true;
            }
        };
    }
    Ok(())
}

File: tests/workflow/consumer/typescript/src/test/test_request_groupb_groupc_structb.ts
// tslint:disable: variable-name
import { Protocol, GroupBGroupCStructB } from "../consumer/index";
import { Consumer } from "../consumer";

export function test(consumer: Consumer): Promise<void> {
	return new Promise((resolve, reject) => {
		let case_GroupBGroupCStructA = false;
		let case_StructB = false;
		let case_StructC = false;
		let case_StructD = false;
		function send(): Promise<void> {
			return new Promise((res, rej) => {
				const request = new GroupBGroupCStructB(
					Protocol.GroupB.GroupC.StructB.defaults(),
					consumer
				);
				request.send().then((response) => {
					if (response instanceof Protocol.GroupB.GroupC.StructA) {
						if (!case_GroupBGroupCStructA) {
							case_GroupBGroupCStructA = true;
							res(undefined);
						} else {
							rej(
								new Error(
									`Multiple times get: Protocol.GroupB.GroupC.StructA`
								)
							);
						}
					} else if (response instanceof Protocol.StructB) {
						if (!case_StructB) {
							case_StructB = true;
							res(undefined);
						} else {
							rej(
								new Error(
									`Multiple times get: Protocol.StructB`
								)
							);
						}
					} else if (response instanceof Protocol.StructC) {
						if (!case_StructC) {
							case_StructC = true;
							res(undefined);
						} else {
							rej(
								new Error(
									`Multiple times get: Protocol.StructC`
								)
							);
						}
					} else if (response instanceof Protocol.StructD) {
						if (!case_StructD) {
							case_StructD = true;
							res(undefined);
						} else {
							rej(
								new Error(
									`Multiple times get: Protocol.StructD`
								)
							);
						}
					}
				});
			});
		}
		function trigger() {
			send()
				.then(() => {
					if (
						case_GroupBGroupCStructA &&
						case_StructB &&
						case_StructC &&
						case_StructD
					) {
						resolve(undefined);
					} else {
						trigger();
					}
				})
				.catch(reject);
		}
		trigger();
	});
}
